controllers/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -107,3 +120,162 @@ func (r *Reconciler) addNewTierUpdate(ctx context.Context, tier *toolchainv1alph
+	ns, err := configuration.GetWatchNamespace()
+	if err != nil {
+		return false, err
+	}
+
+	// tierTemplate doesn't support TTRs
+	// we set TierTemplate as revisions
+	// TODO this step will be removed once we convert all TierTemplates to TTRs
+	if tierTemplate.Spec.TemplateObjects == nil {
+		val, ok := nsTmplTier.Status.Revisions[tierTemplate.GetName()]
+		if !ok || val != tierTemplate.GetName() {
+			nsTmplTier.Status.Revisions[tierTemplate.GetName()] = tierTemplate.GetName()
+			return true, nil
+		}
+		// nothing to update
+		return false, nil
+	}
+
+	if tierTemplateRevisionName, found := nsTmplTier.Status.Revisions[tierTemplate.GetName()]; found {
+		var tierTemplateRevision toolchainv1alpha1.TierTemplateRevision
+		if err := r.Client.Get(ctx, types.NamespacedName{Namespace: ns, Name: tierTemplateRevisionName}, &tierTemplateRevision); err != nil {
+			// no tierTemplateRevision CR was found,
+			if errors.IsNotFound(err) {
+				// let's create one
+				if err := r.createNewTierTemplateRevision(ctx, nsTmplTier, tierTemplate); err != nil {
+					return false, err
+				}
+				// new ttr created
+				return true, nil
+			} else {
+				// something wrong happened
+				return false, err
+			}
+		}
+		// TODO compare TierTemplate content with TTR content
+		// if the TierTemplate has changes we need to create new TTR
+	} else {
+		// no revision was set for this TierTemplate CR, let's create a TTR for it
+		if err := r.createNewTierTemplateRevision(ctx, nsTmplTier, tierTemplate); err != nil {
+			return false, err
+		}
+		// new ttr created
+		return true, nil
+	}
+	// nothing changed
+	return false, nil
+}
+
+func (r *Reconciler) createNewTierTemplateRevision(ctx context.Context, nsTmplTier *toolchainv1alpha1.NSTemplateTier, tierTemplate toolchainv1alpha1.TierTemplate) error {
+	newTTR, err := r.createTTR(ctx, &tierTemplate)
+	if err != nil {
+		// something went wrong while creating new ttr
+		return err
+	}
+	// update NSTemplateTierStatus with new TTR
+	nsTmplTier.Status.Revisions[tierTemplate.GetName()] = newTTR.GetName()
+	return nil
+}
+
+// getNSTemplateTierRefs returns a list with all the refs from the NSTemplateTier
+func getNSTemplateTierRefs(tmplTier *toolchainv1alpha1.NSTemplateTier) []string {
+	var refs []string
+	for _, ns := range tmplTier.Spec.Namespaces {
+		refs = append(refs, ns.TemplateRef)
+	}
+	if tmplTier.Spec.ClusterResources != nil {
+		refs = append(refs, tmplTier.Spec.ClusterResources.TemplateRef)
+	}
+
+	roles := make([]string, 0, len(tmplTier.Spec.SpaceRoles))
+	for r := range tmplTier.Spec.SpaceRoles {
+		roles = append(roles, r)
+	}
+	for _, r := range roles {
+		refs = append(refs, tmplTier.Spec.SpaceRoles[r].TemplateRef)
+	}
+	return refs
+}
+
+func (r *Reconciler) createTTR(ctx context.Context, tmplTier *toolchainv1alpha1.TierTemplate) (*toolchainv1alpha1.TierTemplateRevision, error) {
+	ttr := NewTTR(tmplTier)
+	if err := controllerutil.SetControllerReference(tmplTier, ttr, r.Scheme); err != nil {
```

Comment:
Correct , we still need to create a controller for cleaning up TTRs not used by any space, see : https://issues.redhat.com/browse/KUBESAW-144

But that it's in addition to setting the owner reference, which will cleanup the all the TTRs generated by a specific TierTemplate. It's also very useful in e2e tests for cleaning up TTRs ( since those are generated by the controller ) and not by the tests.

Reply:
for prod - I think that it's fine to not clean it up until we have the proper mechanism in place
for e2e sets - if we need to execute some additional cleanup, then let's do it in the tests, not in the prod code

You could also technically create ownerRef to NSTemplateTier instead of TierTemplate. We still need to define how the cleanup mechanism should work
1. should it be cleaned when either TierTemplate or NSTemplateTier is removed? TTR CR technically depends on both of them, and I would say that it depends more on the NSTemplateTier than the TierTemplate
2. or should we rather delete it as soon as there is no Space/NSTemplateSet that would rely on the TTR CR

if the second, then setting ownerRef for the TierTemplate is wrong approach. 
How we could calculate the delta for changes when one TierTemplate is removed and replaced by another one? For this purpose, we still need to keep the TTR present.

Reply:
> How we could calculate the delta for changes when one TierTemplate is removed and replaced by another one? For this purpose, we still need to keep the TTR present.

Correct me if I'm wrong but the delta should be calculated before we generate a new TTR, basically comparing the content of the new TierTemplate + NStemplateTier params, with the current TTR, if changed then we generate new TTR and the cleanup controller will remove the old one.

Reply:
> You could also technically create ownerRef to NSTemplateTier instead of TierTemplate. We still need to define how the cleanup mechanism should work

I would go with the 2 option, as briefly described in the design doc. 

I would go with the 1 one since we cannot delete a TTR if if it's still used by a space IMO , we would not be able to compare the delta with a new TierTemplate and NSTemplateTier, but maybe I'm missing something.

Reply:
> Correct me if I'm wrong but the delta should be calculated before we generate a new TTR, basically comparing the content of the new TierTemplate + NStemplateTier params with the current TTR, if changed then we generate new TTR and the cleanup controller will remove the old one.

I was talking about the delta when applying the new version of the tier for the end-user (for a Space that is being updated)

> I would NOT go with the 1 one

fully agree, let's drop the ownerReference then

Reply:
Ok, removed the ownerreference in https://github.com/codeready-toolchain/host-operator/pull/1103/commits/085cc00f7785925ff141221ca70d9b9b049480ab, I'll address the cleanup in the tests for now.

Thanks for the suggestion.